core/typevarcheck.ml

```diff
@@ -206,6 +210,8 @@ and is_positive : TypeVarSet.t -> int -> datatype -> bool =
         | `Alias (_, t) -> isp t
         | `Application (_, ts) ->
             List.exists (is_positive_type_arg bound_vars var) ts
+        | `RecursiveApplication { r_args ; _ } ->
```

Comment:
I'm not sure if this is correct. When determining whether some type `Foo(a)` is positive w.r.t. `a`, we should check whether `a` is positive w.r.t. the type that results from unrolling Foo and substituting `a` in. At least that's what the old code would have done if `Foo` is  an alias for a recursive variant type. But that's something our resident type experts have a better grasp on than me...

Reply:
Would it suffice to copy the behaviour of `Recursive` `MetaTypeVar`s here? (Additionally, I think this also means that the definition of `is_guarded_type` and `is_positive_type` should also change accordingly).

Additionally, does anyone have any good literature references on positive / negative types? I feel I could benefit from reading up on this a bit more.

Reply:
Yes mimicking the current behavior for a `Recursive` `MetaTypeVar` should do the trick

Reply:
This should now be resolved in 15db7100b0015db7138f7afca86ea5a52ffa0443.